### PR TITLE
[WIP] Add squash and archive history functionality

### PR DESCRIPTION
**Pull request by AnthonyMl**
co-dependent on https://github.com/bdionne/revision-history/pull/4 and https://github.com/bdionne/protege-server/pull/7
TODO:
- remove project names from history and snapshot files

### CODE REVIEW

**Review by rgrinberg — APPROVED** [review]
LGTM. Seems simple enough

**Comment by bdionne** [comment]
this looks good. As commented on https://github.com/bdionne/revision-history/pull/4, I'd like to wait and merge all three repos at the same time, as part of a new build

File: src/main/java/org/protege/editor/owl/client/LocalHttpClient.java [hunk 1]
```diff
@@ -61,10 +70,16 @@
...
 
+	private static final String SNAPSHOT_CHECKSUM = "-checksum";
+
 	private final String serverAddress;
 
 	private final OkHttpClient httpClient;
 
+	private final ListeningExecutorService service = MoreExecutors.listeningDecorator(Executors.newSingleThreadExecutor());
+
+	private ProgressDialog dlg = new ProgressDialog();
```
Comment (rgrinberg) [inline src/main/java/org/protege/editor/owl/client/LocalHttpClient.java]: Can we scope these instance variables to the methods where they're being used?

The ProgressDialog initialization fails for some unit tests for example.

**Review by rgrinberg — CHANGES_REQUESTED** [review]

File: src/main/java/org/protege/editor/owl/client/LocalHttpClient.java [hunk 2]
```diff
@@ -873,15 +874,31 @@ public boolean codeIsLessThan(String lower, String upper) {
...
+		} catch (ClientRequestException e) {
+			e.printStackTrace();
+			return false;
+		}
     }
+
+    public ServerStatus getServerStatus() throws ClientRequestException {
+		Response response = null;
+		try {
+			response = get(SERVER_STATUS);
```
Comment (bdionne) [inline src/main/java/org/protege/editor/owl/client/LocalHttpClient.java]: This constant is not defined

Comment (AnthonyMl) [reply src/main/java/org/protege/editor/owl/client/LocalHttpClient.java]: Its in
src/main/java/org/protege/editor/owl/server/http/ServerEndpoints.java
in the protege-server pr

Comment (bdionne) [reply src/main/java/org/protege/editor/owl/client/LocalHttpClient.java]: oh ok, I probably haven't pulled latest from there

**Review by rgrinberg — APPROVED** [review]

**Comment by rgrinberg** [comment]
@AnthonyMl Add some `@Nonnull` annotations on the projectId's methods that you've just added/modified.

File: src/main/java/org/protege/editor/owl/client/LocalHttpClient.java [hunk 3]
```diff
@@ -467,11 +467,12 @@ public SnapShot getSnapShot(@Nonnull ProjectId projectId) throws LoginTimeoutExc
 		ClientRequestException {
 		String requestUrl = PROJECT_SNAPSHOT + "?projectid=" + projectId.get();
 		Response response = get(requestUrl); // send request to server
-		return retrieveDocumentSnapshotFromServerResponse(response);
+		return retrieveDocumentSnapshotFromServerResponse(response, projectId);
 	}
 
-	private SnapShot retrieveDocumentSnapshotFromServerResponse(Response response)
+	private SnapShot retrieveDocumentSnapshotFromServerResponse(Response response, @Nonnull ProjectId projectId)
```
Comment (bdionne) [inline src/main/java/org/protege/editor/owl/client/LocalHttpClient.java]: this seems odd, it's declared nonnull then there's a check for a null, and either way the variable is never used?

Comment (rgrinberg) [reply src/main/java/org/protege/editor/owl/client/LocalHttpClient.java]: The boilerplate is there for 2 reasons:

1) Nonnull allows intellij to try and look for places in the code where a null parameter might be passed in.
2) The null check will give a decent error message if it actually occurs at runtime. (These checks can be removed once we're confident that we really aren't passing nulls)

And we are actually using the projectId. It's just that we were erroneously getting it from the instance variable where it was null.

Comment (bdionne) [reply src/main/java/org/protege/editor/owl/client/LocalHttpClient.java]: ? - I think it was an instance variable by design, passing them in is a change

**Comment by rgrinberg** [comment]
The last commit gets rid of that `projectId` local variable entirely. One place where I'd like to see review are those `isWorkflowManager` and `isWorflowModeler` calls. Right now when those client calls fail I'm simply converting them to exceptions. Where the old behavior was to return false. Seems like this is slightly safer. But would like some confirmation there.

**Comment by bdionne** [comment]
-0 on this last commit.

**Comment by rgrinberg** [comment]
I addressed the objections to the last commit. The semantics for modeler/manager checking should now be identical. It still seems worrisome to me that we can swallow any client exception in those calls but at least it's safe for now.

Note that this change also requires tweaking nci-edit-tab as well. So I will make that change as well.

File: src/main/java/org/protege/editor/owl/client/LocalHttpClient.java [hunk 4]
```diff
@@ -828,12 +828,11 @@ public void putEVSHistory(String code, String name, String operation, String ref
...
 
-	public void genConceptHistory()
+	public void genConceptHistory(@Nonnull ProjectId projectId)
 		throws LoginTimeoutException, AuthorizationException, ClientRequestException {
+		if (projectId == null) throw new IllegalArgumentException("projectId cannot be null");
 		try {
-
-			get(GEN_CON_HIST + "?projectid=" + projectId.get());
-
+			get(GEN_CON_HIST + "?projectid=" + projectId);
```
Comment (AnthonyMl) [inline src/main/java/org/protege/editor/owl/client/LocalHttpClient.java]: We need the ```.get()``` here.
toString is different than just the internal id.

**Comment by rgrinberg** [comment]
closed in favor of #20

**Pull request closed without merging**